File: textMSA-frontend/src/api/service.ts
/**
 * Service API (TypeScript)
 * 服务管理和执行相关 API
 */
import request from './request'

// ==================== 类型定义 ====================

/**
 * Service执行状态
 */
export type ExecutionStatus = 'pending' | 'running' | 'completed' | 'failed'


/**
 * 参数定义Schema
 */
export interface ParameterSchema {
  [paramName: string]: {
    type: 'string' | 'number' | 'integer' | 'boolean' | 'array' | 'object' | 'enum' | 'continuous' | 'discrete'
    default?: any
    default_value?: any
    enum?: any[]
    enum_values?: any[]
    min?: number
    max?: number
    min_value?: number
    max_value?: number
    description?: string
    items?: ParameterSchema
  }
}

/**
 * Service信息
 */
/**
 * 输出结果项类型
 */
export type ServiceOutputItemType = 'file' | 'text'

/**
 * 接受的文件类型配置
 * 定义单个文件名的文件类型要求
 */
export interface AcceptedFileConfig {
  /** 文件类型ID列表（支持多个类型） */
  file_type_ids: string[]
  /** 文件描述 */
  description: string
}

/**
 * 接受的文件类型映射
 * key: 文件名（如 "input_data.h5ad"）
 * value: 该文件名的文件类型配置
 */
export interface AcceptedFiles {
  [filename: string]: AcceptedFileConfig
}

/**
 * 文件输出项
 */
export interface FileOutputItem {
  type: 'file'
  filename: string
  description: string
  /** 文件类型ID（必需字段） */
  file_type_id: string
}

/**
 * 文本信息输出项
 */
export interface TextOutputItem {
  type: 'text'
  filename: string
  description: string
}

/**
 * 输出结果配置
 */
export interface ServiceOutputConfig {
  items: Array<FileOutputItem | TextOutputItem>
  collection_description?: string
}

/**
 * Service信息
 */
export interface Service {
  service_id: string
  name: string
  description?: string
  version: string
  baseurl: string
  service_suffix: string
  download_suffix?: string
  parameter_template: Record<string, any>
  parameter_schema?: ParameterSchema
  output_template?: Record<string, any>
  accepted_files?: AcceptedFiles
  output_config?: ServiceOutputConfig
  visibility?: 'private' | 'public' | 'system'
  created_at?: string
  updated_at?: string
  created_by?: string
}

/**
 * Service列表响应
 */
export interface ServiceListResponse {
  services: Service[]
  total: number
}

/**
 * 创建Service请求
 */
export interface ServiceCreateRequest {
  name: string
  description?: string
  version?: string
  baseurl: string
  service_suffix: string
  download_suffix?: string
  parameter_template?: Record<string, any>
  parameter_schema?: ParameterSchema
  accepted_files?: AcceptedFiles
  output_config?: ServiceOutputConfig
  visibility?: 'private' | 'public' | 'system'
}

/**
 * 请求配置（用于向后兼容，保留原有字段）
 */
export type RequestConfig = Record<string, any>

/**
 * 更新Service请求
 */
export interface ServiceUpdateRequest {
  name?: string
  description?: string
  version?: string
  baseurl?: string
  service_suffix?: string
  download_suffix?: string
  visibility?: 'private' | 'public'
  request_config?: RequestConfig
  parameter_template?: Record<string, any>
  parameter_schema?: ParameterSchema
  output_template?: Record<string, any>
  accepted_files?: AcceptedFiles
  output_config?: ServiceOutputConfig
}

/**
 * Service执行请求（与后端对齐）
 */
export interface ServiceExecuteRequest {
  input_file_ids: string[]
  parameters?: Record<string, any>
  project_id?: string
}

/**
 * Service执行响应
 */
export interface ServiceExecution {
  execution_id: string
  service_id: string
  service_name?: string
  user_id: string
  input_file_ids: string[]
  output_file_ids?: string[]
  status: ExecutionStatus
  parameters: Record<string, any>
  response_data?: Record<string, any>
  error_message?: string
  created_at?: string
  started_at?: string
  completed_at?: string
  duration_seconds?: number
}

/**
 * Service执行列表响应
 */
export interface ServiceExecutionListResponse {
  executions: ServiceExecution[]
  total: number
}

// ==================== 类型守卫函数 ====================

/**
 * 检查文件输出项是否包含 file_type_id
 */
export function isFileOutputItemComplete(item: FileOutputItem | TextOutputItem): item is FileOutputItem {
  if (item.type === 'file') {
    return 'file_type_id' in item && typeof item.file_type_id === 'string' && item.file_type_id.length > 0
  }
  return false
}

/**
 * 检查服务是否有 accepted_files 配置
 */
export function hasAcceptedFiles(service: Service): boolean {
  return service.accepted_files !== undefined && 
         service.accepted_files !== null && 
         Object.keys(service.accepted_files).length > 0
}

// ==================== API 函数 ====================

// 服务列表缓存（30秒 TTL）
const SERVICE_LIST_CACHE_TTL = 30000
const serviceListCache = new Map<string, { data: ServiceListResponse; fetchedAt: number }>()

function serviceListCacheKey(visibility?: string, projectId?: string, skip?: number, limit?: number): string {
  return `${visibility || ''}|${projectId || ''}|${skip || 0}|${limit || 100}`
}

/**
 * 获取Service列表（带缓存）
 * @param visibility - 可见性过滤（可选）
 * @param projectId - 项目ID（可选）
 * @param skip - 跳过数量（默认0）
 * @param limit - 返回数量（默认100）
 * @param force - 强制刷新（默认false）
 */
export async function getServiceList(
  visibility?: 'private' | 'public' | 'system',
  projectId?: string,
  skip: number = 0,
  limit: number = 100,
  force: boolean = false
): Promise<ServiceListResponse> {
  const cacheKey = serviceListCacheKey(visibility, projectId, skip, limit)

  if (!force) {
    const cached = serviceListCache.get(cacheKey)
    if (cached && (Date.now() - cached.fetchedAt) < SERVICE_LIST_CACHE_TTL) {
      return cached.data
    }
  }

  const params: any = { skip, limit }
  if (visibility) params.visibility = visibility
  if (projectId) params.project_id = projectId
  
  const result = await request({
    url: '/service/',
    method: 'GET',
    params
  }) as ServiceListResponse

  serviceListCache.set(cacheKey, { data: result, fetchedAt: Date.now() })
  return result
}

/**
 * 清除服务列表缓存
 */
export function clearServiceListCache() {
  serviceListCache.clear()
}

/**
 * 获取Service详情
 * @param serviceId - Service ID
 * 
 * @returns Service对象，包含以下新字段：
 * - accepted_files: 服务接受的输入文件类型（如果配置）
 * - output_config.items 中的文件项包含 file_type_id 字段
 * 
 * @remarks
 * 如果服务是旧数据（没有新字段），响应中可能不包含这些字段。
 * 前端应该优雅处理这种情况。
 */
export async function getService(serviceId: string): Promise<Service> {
  return request({
    url: `/service/${serviceId}`,
    method: 'GET'
  }) as Promise<Service>
}

/**
 * 获取Service详情（别名，用于兼容）
 */
export const getServiceDetail = getService

/**
 * 创建Service
 * @param serviceData - Service创建数据
 * 
 * @remarks
 * 新字段说明：
 * - accepted_files: 可选，定义服务接受的输入文件类型
 * - output_config.items 中的文件项必须包含 file_type_id 字段
 * 
 * @example
 * ```typescript
 * const service = await createService({
 *   name: "服务名称",
 *   baseurl: "https://api.example.com",
 *   service_suffix: "/api/service",
 *   accepted_files: {
 *     "input_data.h5ad": {
 *       file_type_ids: ["preprocessed_single_cell_data"],
 *       description: "输入的单细胞数据文件"
 *     }
 *   },
 *   output_config: {
 *     items: [
 *       {
 *         type: "file",
 *         filename: "output.h5ad",
 *         description: "输出文件",
 *         file_type_id: "preprocessed_single_cell_data"
 *       }
 *     ]
 *   }
 * })
 * ```
 */
export async function createService(serviceData: ServiceCreateRequest): Promise<Service> {
  return request({
    url: '/service/',
    method: 'POST',
    data: serviceData
  }) as Promise<Service>
}

/**
 * 更新Service
 * @param serviceId - Service ID
 * @param updateData - 更新数据
 * 
 * @remarks
 * 可以更新以下新字段：
 * - accepted_files: 更新服务接受的输入文件类型
 * - output_config: 更新输出配置（文件项必须包含 file_type_id）
 * 
 * @example
 * ```typescript
 * await updateService(serviceId, {
 *   accepted_files: {
 *     "input_data.h5ad": {
 *       file_type_ids: ["preprocessed_single_cell_data"],
 *       description: "输入的单细胞数据文件"
 *     }
 *   }
 * })
 * ```
 */
export async function updateService(
  serviceId: string,
  updateData: ServiceUpdateRequest
): Promise<Service> {
  return request({
    url: `/service/${serviceId}`,
    method: 'PUT',
    data: updateData
  }) as Promise<Service>
}

/**
 * 删除Service
 * @param serviceId - Service ID
 */
export async function deleteService(serviceId: string): Promise<{ service_id: string }> {
  return request({
    url: `/service/${serviceId}`,
    method: 'DELETE'
  }) as Promise<{ service_id: string }>
}

/**
 * 执行Service
 * @param serviceId - Service ID
 * @param executeData - 执行数据
 * 
 * @remarks
 * 后端会验证输入文件的 file_type_id 是否符合服务的 accepted_files 要求。
 * 如果文件类型不匹配，会返回 400 错误。
 * 
 * @throws {Error} 如果文件类型不匹配，错误信息会明确指出哪个文件的类型不匹配
 * 
 * @example
 * ```typescript
 * try {
 *   const execution = await executeService(serviceId, {
 *     input_file_ids: ["file_id_1", "file_id_2"],
 *     parameters: { ... }
 *   })
 * } catch (error) {
 *   // 处理文件类型不匹配错误
 *   if (error.message.includes("文件类型")) {
 *     // 显示用户友好的错误提示
 *   }
 * }
 * ```
 */
export async function executeService(
  serviceId: string,
  executeData: ServiceExecuteRequest
): Promise<ServiceExecution> {
  return request({
    url: `/service/${serviceId}/execute`,
    method: 'POST',
    data: executeData
  }) as Promise<ServiceExecution>
}

/**
 * 获取执行记录详情
 * @param executionId - 执行ID
 */
export async function getExecution(executionId: string): Promise<ServiceExecution> {
  return request({
    url: `/service/executions/${executionId}`,
    method: 'GET'
  }) as Promise<ServiceExecution>
}

/**
 * 获取执行记录列表
 * @param serviceId - Service ID过滤（可选）
 * @param userId - 用户ID过滤（可选）
 * @param status - 状态过滤（可选）
 * @param skip - 跳过数量（可选，默认0）
 * @param limit - 返回数量（可选，默认100）
 */
export async function getExecutionList(
  serviceId?: string,
  userId?: string,
  status?: ExecutionStatus,
  project?: string,
  skip: number = 0,
  limit: number = 100
): Promise<ServiceExecutionListResponse> {
  const params: any = { skip, limit }
  if (serviceId) params.service_id = serviceId
  if (userId) params.user_id = userId
  if (status) params.status = status
  if (project) params.project = project
  
  return request({
    url: '/service/executions/',
    method: 'GET',
    params
  }) as Promise<ServiceExecutionListResponse>
}

File: textMSA/textmsa/services/api/routers/service.py
"""
Service API路由
"""
from fastapi import APIRouter, HTTPException, status, Request, Query
from fastapi.responses import JSONResponse
from typing import Optional

from textmsa.logging_config import get_logger
from textmsa.services.auth.auth_service import get_current_user_from_header
from textmsa.services.service.service_service import get_service_service
from textmsa.services.api.schemas import (
    ServiceCreateRequest,
    ServiceUpdateRequest,
    ServiceResponse,
    ServiceListResponse,
    ServiceExecuteRequest,
    ServiceExecuteResponse,
    ServiceExecutionResponse,
    ServiceExecutionListResponse
)

logger = get_logger(__name__)

# 创建路由
service_router = APIRouter(prefix="/api/service", tags=["Service管理"])


@service_router.post("/", response_model=ServiceResponse)
async def create_service(
    request: Request,
    service_data: ServiceCreateRequest
):
    """
    创建Service
    
    - **name**: Service名称
    - **description**: Service描述（可选）
    - **version**: Service版本（可选，默认1.0.0）
    - **baseurl**: 基础URL（IP和端口号，如 http://192.168.1.1:8080）
    - **service_suffix**: Service后缀（与baseurl拼接后访问对应的service，如 /api/service）
    - **download_suffix**: Download后缀（可选，如果输出存在文件则需要配置，用于下载生成文件，如 /api/download）
    - **tags**: Service标签集合（可选）
    - **requestConfig**: 请求配置（可选）
    - **parameterTemplate**: 参数模板（可选）
    - **visibility**: Service权限（可选，默认private；可选：private/public/system）
    
    注意：
    - service_id 如果未提供，系统会自动生成 UUID
    - 默认visibility为PRIVATE（仅创建者可见）
    
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户
        user_info = await get_current_user_from_header(request=request)
        
        user_id = user_info.get("user_id") if user_info else None
        
        # 创建Service
        service_service = get_service_service()
        service = service_service.create_service(service_data.model_dump(), user_id=user_id)
        
        return JSONResponse(
            status_code=status.HTTP_201_CREATED,
            content={
                "code": 201,
                "message": "Service创建成功",
                "data": service
            }
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"创建Service失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"创建Service失败: {str(e)}"
        )


@service_router.get("/list", response_model=ServiceListResponse)
@service_router.get("", response_model=ServiceListResponse)
@service_router.get("/", response_model=ServiceListResponse)
async def list_services(
    request: Request,
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(100, ge=1, le=1000, description="返回数量"),
    visibility: Optional[str] = Query(None, alias="visibility", description="Service权限（可选，默认private；可选：private/public/system）"),
    project_id: Optional[str] = Query(None, alias="projectId", description="项目ID（用于项目过滤）"),
    input_file_type_id: Optional[str] = Query(None, alias="inputFileTypeId", description="输入文件类型ID（用于筛选接受该文件类型的服务）")
):
    """
    获取Service列表
    
    - **skip**: 跳过数量（可选，默认0）
    - **limit**: 返回数量（可选，默认100，最大1000）
    - **visibility**: Service权限（可选，默认private；可选：private/public/system）
    - **projectId**: 项目ID（可选，用于项目过滤）
    - **inputFileTypeId**: 输入文件类型ID（可选，用于筛选接受该文件类型的服务）
    
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户（可选，用于权限控制）
        user_info = await get_current_user_from_header(request=request)
        user_id = user_info.get("user_id") if user_info else None
        
        # 获取Service列表（根据权限过滤）
        service_service = get_service_service()
        logger.info(f"获取Service列表，visibility={visibility}, project_id={project_id}, input_file_type_id={input_file_type_id}")
        result = service_service.list_services(
            skip=skip,
            limit=limit,
            user_id=user_id,
            visibility_filter=visibility,
            project_id=project_id,
            input_file_type_id=input_file_type_id
        )
        
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "code": 200,
                "message": "success",
                "data": result
            }
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取Service列表失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"获取Service列表失败: {str(e)}"
        )


@service_router.put("/{service_id}", response_model=ServiceResponse)
async def update_service(
    service_id: str,
    request: Request,
    update_data: ServiceUpdateRequest
):
    """
    更新Service信息
    
    - **service_id**: Service ID（路径参数）
    - **update_data**: 更新数据（只包含要更新的字段）
    
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户（可选，用于权限控制）
        # user_info = await get_current_user_from_header(request=request)
        
        
        # 更新Service
        service_service = get_service_service()
        service = service_service.update_service(service_id, update_data.model_dump(exclude_unset=True))
        
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "code": 200,
                "message": "Task更新成功",
                "data": service
            }
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"更新Service失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"更新Service失败: {str(e)}"
        )


@service_router.delete("/{service_id}")
async def delete_service(
    service_id: str,
    request: Request
):
    """
    删除Service
    
    - **service_id**: Service ID
    
    注意：如果Task存在执行记录，则无法删除
    
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户（可选，用于权限控制）
        user_info = await get_current_user_from_header(request=request)
        
        # 删除Service
        service_service = get_service_service()
        success = service_service.delete_service(service_id)
        
        if success:
            return JSONResponse(
                status_code=status.HTTP_200_OK,
                content={
                    "code": 200,
                    "message": "Task删除成功",
                    "data": {"service_id": service_id}
                }
            )
        else:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Task删除失败"
            )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"删除Service失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"删除Service失败: {str(e)}"
        )


@service_router.get("/{service_id}", response_model=ServiceResponse)
async def get_service_info(
    service_id: str,
    request: Request
):
    """
    获取Service信息
    
    - **service_id**: Service ID
    
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户（可选，用于权限控制）
        user_info = await get_current_user_from_header(request=request)
        user_id = user_info.get("user_id") if user_info else None
        
        # 获取Service信息（带权限检查）
        service_service = get_service_service()
        service_info = service_service.get_service(service_id, user_id=user_id)
        
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "code": 200,
                "message": "success",
                "data": service_info
            }
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取Service信息失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"获取Service信息失败: {str(e)}"
        )


@service_router.get("/queue/status")
async def get_queue_status(request: Request):
    """
    获取服务执行队列状态
    
    返回当前运行数、排队数和最大并发数
    """
    try:
        service_service = get_service_service()
        queue_status = service_service.get_queue_status()
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "code": 200,
                "message": "success",
                "data": queue_status
            }
        )
    except Exception as e:
        logger.error(f"获取队列状态失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"获取队列状态失败: {str(e)}"
        )


@service_router.post("/{service_id}/execute", response_model=ServiceExecuteResponse)
async def execute_service(
    service_id: str,
    request: Request,
    execute_data: ServiceExecuteRequest
):
    """
    执行Service
    
    - **service_id**: Service ID（路径参数）
    - **input_file_ids**: 输入文件ID列表（支持多文件输入）
    - **parameters**: 执行参数（可选，覆盖parameterTemplate）
        
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户
        user_info = await get_current_user_from_header(request=request)
        
        if not user_info:
            logger.warning(f"[API] 用户未授权 - service_id: {service_id}")
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="未授权，请先登录"
            )
        
        user_id = user_info["user_id"]
        
        # 执行Service
        service_service = get_service_service()
        result = service_service.execute_service(
            service_id=service_id,
            input_file_ids=execute_data.input_file_ids,
            user_id=user_id,
            parameters=execute_data.parameters,
            project_id=execute_data.project_id
        )
        
        execution_id = result.get('execution_id') if isinstance(result, dict) else 'unknown'
        logger.info(f"[API] Service执行请求已提交 - execution_id: {execution_id}, service_id: {service_id}, status: {result.get('status')}")
        
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "code": 200,
                "message": "Task执行成功",
                "data": result
            }
        )
    
    except HTTPException as http_ex:
        logger.error(f"[API] HTTP异常 - service_id: {service_id}, status_code: {http_ex.status_code}, detail: {http_ex.detail}")
        raise
    except Exception as e:
        logger.error(f"[API] 执行Service失败 - service_id: {service_id}, error: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"执行Service失败: {str(e)}"
        )


@service_router.get("/executions/{execution_id}", response_model=ServiceExecutionResponse)
async def get_execution(
    execution_id: str,
    request: Request
):
    """
    获取Task执行记录
    
    - **execution_id**: 执行ID
    
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户（可选，用于权限控制）
        user_info = await get_current_user_from_header(request=request)
        
        # 获取执行记录
        service_service = get_service_service()
        execution = service_service.get_execution(execution_id)
        
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "code": 200,
                "message": "success",
                "data": execution
            }
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取执行记录失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"获取执行记录失败: {str(e)}"
        )


@service_router.get("/executions/", response_model=ServiceExecutionListResponse)
async def list_executions(
    request: Request,
    service_id: Optional[str] = Query(None, description="Service ID过滤"),
    user_id: Optional[str] = Query(None, description="用户ID过滤"),
    status_filter: Optional[str] = Query(None, description="状态过滤", alias="status"),
    project_id: Optional[str] = Query(None, description="Project ID过滤", alias="project"),
    skip: int = Query(0, ge=0, description="跳过数量"),
    limit: int = Query(100, ge=1, le=1000, description="返回数量")
):
    """
    获取Task执行记录列表
    
    - **service_id**: Service ID过滤（可选）
    - **user_id**: 用户ID过滤（可选）
    - **status**: 状态过滤（可选）
    - **project**: Project ID过滤（可选）
    - **skip**: 跳过数量（可选，默认0）
    - **limit**: 返回数量（可选，默认100，最大1000）
    
    需要提供Authorization头或token头
    """
    try:
        # 获取当前用户（可选，用于权限控制）
        user_info = await get_current_user_from_header(request=request)
        
        # 如果用户已登录，默认只返回该用户的执行记录
        if user_info and not user_id:
            user_id = user_info.get("user_id")
        
        # 获取执行记录列表
        service_service = get_service_service()
        result = service_service.list_executions(
            service_id=service_id,
            user_id=user_id,
            status_filter=status_filter,
            project_id=project_id,
            skip=skip,
            limit=limit
        )
        
        return JSONResponse(
            status_code=status.HTTP_200_OK,
            content={
                "code": 200,
                "message": "success",
                "data": result
            }
        )
    
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取执行记录列表失败: {e}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"获取执行记录列表失败: {str(e)}"
        )
